File: src/main/java/com/demonwav/mcdev/platform/forge/inspections/sideonly/RemoveAnnotationInspectionGadgetsFix.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.forge.inspections.sideonly;

import com.demonwav.mcdev.platform.forge.util.ForgeConstants;

import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiModifierList;
import com.intellij.psi.PsiModifierListOwner;
import com.siyeh.ig.InspectionGadgetsFix;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class RemoveAnnotationInspectionGadgetsFix extends InspectionGadgetsFix {

    @Nullable
    public abstract PsiModifierListOwner getListOwner();

    @Override
    protected void doFix(Project project, ProblemDescriptor descriptor) {
        final PsiModifierListOwner owner = getListOwner();
        if (owner == null) {
            return;
        }

        final PsiModifierList list = owner.getModifierList();
        if (list == null) {
            return;
        }

        final PsiAnnotation annotation = list.findAnnotation(ForgeConstants.SIDE_ONLY_ANNOTATION);
        if (annotation == null) {
            return;
        }

        annotation.delete();
    }

    @Nls
    @NotNull
    @Override
    public abstract String getName();

    @Nls
    @NotNull
    @Override
    public String getFamilyName() {
        return getName();
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/forge/inspections/sideonly/RemoveAnnotationInspectionGadgetsFix.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.forge.inspections.sideonly

import com.demonwav.mcdev.platform.forge.util.ForgeConstants
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiModifierListOwner
import com.intellij.structuralsearch.plugin.util.SmartPsiPointer
import com.siyeh.ig.InspectionGadgetsFix
import org.jetbrains.annotations.Nls

class RemoveAnnotationInspectionGadgetsFix(element: PsiModifierListOwner, private val name: String) :
    InspectionGadgetsFix() {

    private val pointer: SmartPsiPointer = SmartPsiPointer(element)

    override fun doFix(project: Project, descriptor: ProblemDescriptor) {
        val owner = pointer.element as? PsiModifierListOwner ?: return
        val list = owner.modifierList ?: return
        val annotation = list.findAnnotation(ForgeConstants.SIDE_ONLY_ANNOTATION) ?: return

        annotation.delete()
    }

    @Nls
    override fun getName() = name

    @Nls
    override fun getFamilyName(): String {
        return name
    }
}
